### PR TITLE
Refactor Review Moderation logic part 2: handle queuing

### LINKED ISSUE #5647: [Reviews] Remove the tight coupling between ReviewDetailFragment and ReviewListFragment during reviews moderation
Currently product reviews moderation is handled using `EventBus`, and the logic is causing a very tight coupling between `ReviewDetailFragment` and `ReviewListFragment`, which makes reusing the `ReviewDetailFragment` in a different navigation flow impossible (~~This is blocking https://github.com/woocommerce/woocommerce-android/issues/5210~~)

We should think about a more robust architecture for handling this, while allowing the logic we have now, ideally the changes would be:
1. Remove the usage of `EventBus` from the communication between the screens.
2. Make `ReviewDetailFragment` independent from having `ReviewListFragment` on the backstack.
3. Keep the logic of async moderation: show a Snackbar on the parent screen with `undo` button.
4. Nice to have: migrate `newUpdateProductReviewStatusAction` to a suspendable function.

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #5647 

### Description
This PR improves the handling of queuing moderation requests, this was broken in the past implementation, and in the previous PR #6037 too.
Check the videos below to see how this is improved:

- On the past implementation: requests were lost if a request was still pending.
- After #6037: there was a UI glitch where the statuses are not correct, and the undo snack bar wasn't shown until the first previous request is completed.

| Scenario | Before | After PR 1 | After PR 2|
| ---- | ---- | ---- | ---- |
| Approval change |  |  |  |
| Mark as spam |  |  |  |

### Testing instructions
I don't think it makes sense to give specific test instructions for this PR, just try to break it, and give your feedback on how it behaves.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt [hunk 1]
```diff
@@ -9,10 +10,7 @@ import com.woocommerce.android.tools.SelectedSite
 import com.woocommerce.android.util.WooLog
 import com.woocommerce.android.util.WooLog.T
 import kotlinx.coroutines.*
-import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.MutableSharedFlow
-import kotlinx.coroutines.flow.asSharedFlow
-import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.*
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:15]: 💡 We might want to consider adding a comment on top of this class and explain it's purpose. The variable names and implementation assume the reader understands the purpose of the class and what problems it solves. 
Since I didn't understand the purpose and the problems, it took me over an hour to fully understand what is going on. Perhaps a couple sentences what the class does + a couple paragraphs with examples of UI glitches the class solves might help future readers. Wdyt?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt [hunk 2]
```diff
@@ -24,56 +22,80 @@ class ReviewModerationHandler @Inject constructor(
...
+
+    private val _skipDelayTrigger = MutableSharedFlow<Unit>()
 
-    private var moderationRequestJob: Job? = null
+    private var pendingJobs = mutableMapOf<Long, Job>()
 
     init {
         appCoroutineScope.launch {
-            _queue.collect {
+            // Queue requests
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt]: ❓ I'm still learning some tips and tricks with Flows and Coroutines, so take this comment with a grain of salt, I might be missing something vital.

It seems that `queue requests` launches a new coroutine and collects the flow using `launchIn` operator. However, the `Process requests` collects it within this coroutine which is launched at the top of this method.
Is there any reason why we combine usage of these two approaches?

Would there be any change in behavior if the method looked like this:
```
init {
   // Queue requests
   _queue.onEach {...}.launchIn(appCoroutineScope)
   // Process requests
   _queue.onEach {...}.launchIn(appCoroutineScope)
}
```
Or it could be the other way round, both methods would use `launch { ...collect {} })`. I basically just don't understand if there is any reason to combine the two approaches. 

Wdyt?

P.S. We might want to consider removing `// Queue requests and // Process requests` comments and extracting the methods into `queueIncomingRequests` and `processIncomingRequests` methods or something in that sense, wdyt?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt [hunk 3]
```diff
@@ -78,12 +79,34 @@ class ReviewModerationHandlerTests : BaseUnitTest() {
...
             handler.postModerationRequest(review, HOLD)
             advanceTimeBy(ReviewModerationHandler.UNDO_DELAY)
         }
 
-        assertThat(latestStatus?.actionStatus).isEqualTo(ActionStatus.SUCCESS)
+        assertThat(latestStatus.actionStatus).isEqualTo(ActionStatus.SUCCESS)
+    }
+
+    @Test
+    fun `when review status update succeeds, then keep status for a given time`() = testBlocking {
```
Comment (malinajirka) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt]: ❓ It took me a while to understand the purpose of this test. Would it make sense to rephrase this to something like
`when review status update succeeds, then status removed after a given time`
It feels like that's what the test actually verifies -> that it gets removed, not that it's present until the time passes. 

Or perhaps even clearer would be to include the reason why we keep the status for a given time - to improve documentation value of the test. Perhaps something like `when review status update succeeds, then status removed after the client had chance to reload data from db`.

Same thing applies to `when review status update fails, then keep status for a given time`.

Wdyt?

**Review by malinajirka — APPROVED** [review]
Thanks @hichamboushaba ! Great job fixing the tricky ui glitches!

I've left a few in-code comments, but none of them is blocking.

I also noticed one more UI glitch, but I'm not sure if it's related to any of the changes in this PR series. 
1. Open a review
2. (Un)approve it
3. Open the same review again and notice the ui glitch on the snackbar

https://user-images.githubusercontent.com/2261188/159842490-3a21d702-1fd2-45ca-b85a-b04ce9d51b85.mp4

**Comment by malinajirka** [comment]
I think I found one more edge-case glitch
1. (un)approve a review
2. Open the same review again and (un)approve it
3. Notice when there are two pending requests for the same review, the UI shows the state of the first request not the latest

Since this is an edge case and doesn't result in data loss, consider whether it's even worth fixing.

https://user-images.githubusercontent.com/2261188/159844180-fc0edb95-8138-4c62-bfbc-de099709e99f.mp4

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:15]: > Since I didn't understand the purpose and the problems

Just to understand what's the reason of this, after reading the PR description, did you check the original issue #5647 and the previous PR #6037? and did they help clarify the purpose of the class or not?

> a couple paragraphs with examples of UI glitches the class solves might help future readers

The goal of the class is not a fix the UI glitches, its purpose is to make the review moderation handling reusable, in order to be able to add it to the screen in product details.
The fix for UI glitches is just an additional task that I wanted to tackle while working on this 🙂

I added a comment in 092cd98, please let me know if you have a better suggestion 🙂.

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:15]: > Just to understand what's the reason of this, after reading the PR description, did you check the original issue https://github.com/woocommerce/woocommerce-android/issues/5647 and the previous PR https://github.com/woocommerce/woocommerce-android/pull/6037? and did they help clarify the purpose of the class or not?

Yes, I checked them and I understood the reason why we are doing this refactoring, however, I didn't understood how the whole review moderation process is working. Eg. We optimisticly change the status -> but not in db, but just the handler -> we wait until the undo snackbar disappears -> we send the request to the server -> we change status to success but keep it in the handler to give the client time to reload the data from the db (if we removed it from the handler on success, the client would use the in-memory cached data which still contain the original model).

> The goal of the class is not a fix the UI glitches, its purpose is to make the review moderation handling reusable, in order to be able to add it to the screen in product details.
The fix for UI glitches is just an additional task that I wanted to tackle while working on this 🙂
I added a comment in https://github.com/woocommerce/woocommerce-android/commit/092cd983c14a788e2a8b6bc1a3399c51de35a11e, please let me know if you have a better suggestion 🙂

Thanks for the comment, that definitely improves it. However, I'd still consider explaining how it works -> something like what I mentioned in the previous paragraph. Wdyt?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:15]: > However, I'd still consider explaining how it works -> something like what I mentioned in the previous paragraph. Wdyt?

Sounds good, I added some details in the commit c8e8b0d, feel free to suggest any more improvements 🙂

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt]: > Is there any reason why we combine usage of these two approaches?

The only reason is that `collect` is a terminal suspendable function, so making it the last function in the launched coroutine to pause while processing requests made sense.
The only difference this will have with both `launch { collect }` and `launchIn` is that those will result in creating an additional child coroutine explicitly, but this doesn't make any real difference.

Anyway, to solve the UI glitches you found out about, I had to update the logic on the enqueuing and processing to make sure it's executed in an ordered manner, as using multiple collectors was not definitive on the order of execution, and resulted in issues when adding multiple requests for the same review, check the commit: 931c5d3

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt]: Done in 3b3e327

**Comment by hichamboushaba** [comment]
Thanks so much @malinajirka for the review, and nice find on the glitches.

> I also noticed one more UI glitch, but I'm not sure if it's related to any of the changes in this PR series.

As you guessed, this is not really related to changes of this PR, as it seems it's a bug with the Snackbar animations/implementation, sometimes it flickers like this when navigating to a new screen. I confirmed this by removing the code to `dismiss` the Snackbar, and it was still happening:

https://user-images.githubusercontent.com/1657201/159914476-eece29b9-5630-46f2-89d5-bcea1a98cb31.mov

I made sure with breakpoints nothing is making changes to the Snackbars.

So to avoid the issue for our case, I updated the logic of `ReviewModerationUi` to make sure the Snackbar is dismissed in `onStop` and not `onViewDestroyed` in the commit edf8451

> I think I found one more edge-case glitch

This should be fixed by the [commit](https://github.com/woocommerce/woocommerce-android/pull/6052/commits/931c5d3d7edb2a69249af1a6832457db4d2d9b25) that updated the logic of enqueuing/processing, as now we make sure to cancel any past requests of the review, before enqueuing the new one.

Ready for another round, and thanks again.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt]: Done in 3b3e327

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt]: > will result in creating an additional child coroutine explicitly, but this doesn't make any real difference.

I believe we wouldn't need the parent/wrapper coroutine in such scenario, so I think the number of coroutines would be exactly the same = 2.

Anyway, your solution in https://github.com/woocommerce/woocommerce-android/commit/931c5d3d7edb2a69249af1a6832457db4d2d9b25 looks great! Thanks!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:15]: It's perfect, thank you so much!

**Review by malinajirka — APPROVED** [review]
Looks great, thanks so much! Great job 👏
